### PR TITLE
At cpp ops map [WIP]

### PR DESCRIPTION
**Pull request by atuzhykov**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [ ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ ] Created tests for any significant new code additions.
- [ ] Relevant tests for your changes are passing.

### CODE REVIEW

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/ops/SDImage.java [hunk 1]
```diff
@@ -254,6 +255,98 @@ public SDVariable hsvToRgb(String name, SDVariable input) {
...
+   * ResizeArea: Anti-aliased resampling with area interpolation. 'antialias' has no effect when used with area interpolation; it always anti-aliases.
+   * ResizeMitchelcubic: Mitchell-Netravali Cubic non-interpolating filter. For synthetic images (especially those lacking proper prefiltering), less ringing than Keys cubic kernel but less sharp.
+   * @return output Output image (NUMERIC type)
+   */
+  public SDVariable imageResize(String name, SDVariable input, SDVariable size,
+      boolean preserveAspectRatio, boolean antialis, ImageResizeMethods ImageResizeMethods) {
+    SDValidation.validateNumerical("imageResize", "input", input);
+    SDValidation.validateNumerical("imageResize", "size", size);
+    SDVariable out =  new org.nd4j.linalg.api.ops.impl.image.ImageResize(sd,input, size, preserveAspectRatio, antialis, ImageResizeMethods).outputVariable();
+    return sd.updateVariableNameAndReference(out, name);
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/ops/SDImage.java:303]: Few small changes to make here:
* Make size argument an integer datatype (i.e., `Input(INT, "size")` not `Input(NUMERIC, "size")` in op definition) - it doesn't make sense to have image size be floating point, etc
* Rename enum: ImageResizeMethods -> ImageResizeMethod (don't forget to manually delete the old generated ImageResizeMethods enum after you regerenate ImageResizeMethod - the code gen tool won't delete the old one)

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/image/ImageResize.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Override
+    public String opName() {
+        return "image_resize";
+    }
+
+
+    public ImageResize(@NonNull SameDiff sameDiff, @NonNull SDVariable in, @NonNull SDVariable size, boolean preserveAspectRatio, boolean antialias, ImageResizeMethods method) {
+        super("image_resize", sameDiff, new SDVariable[]{in, size});
+        Preconditions.checkArgument(in.getArr().shape().length==4,"expected input message in NCHW format i.e [batchSize, channels, height, width]");
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/image/ImageResize.java:29]: Couple of things:
To get the rank (number of dimensions): don't use `INDArray.shape().length` - use `INDArray.rank()` instead

Unfortunately we can't check rank 4 for SDVariable - getArr() will return null, unless `in` in a constant or variable... if it's a placeholder or array type (i.e., output of some op), getArr() will return null because no INDArray exists for that variable yet.
We _can_ check it for the INDArray contsructor - so keep that one (line 36) but delete this one (line 29)

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/image/ImageResize.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package org.nd4j.linalg.api.ops.impl.image;
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/image/ImageResize.java:1]: Don't forget copyright header

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/bp/MergeAvgBp.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package org.nd4j.linalg.api.ops.impl.shape.bp;
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/bp/MergeAvgBp.java:1]: Copyright header
Same thing for MergeMaxBp

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/tensorops/EmbeddingLookup.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Override
+    public String opName() {
+        return "embedding_lookup";
+    }
+
+
+    public EmbeddingLookup(@NonNull SameDiff sameDiff, @NonNull SDVariable in, @NonNull SDVariable indices) {
+        super("embedding_lookup", sameDiff, new SDVariable[]{in, indices});
+        addIArgument(indices.eval().toIntVector());
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/tensorops/EmbeddingLookup.java:44]: We can't do this (`addIArgument(indices.eval().toIntVector());`), for a few reasons:
eval() requires doing forward pass through the entire network up until the SDVariable for indices - it might require a lot of computation.
In many cases, we can't even do it - for example, we might need a placeholder value (yet to be provided by the user) to calculate the indices SDVariable
Finally, even if we _can_ calculate it right now - it doesn't mean that _indices_ won't be different next iteration of training/inference - i.e., the IArgument values will be outdated in future iterations, but will never be updated.

In short: never use `SDVariable.eval()` in any of the operations, for these reasons.

Also, we don't need this - the indices are already an input.
So, instead just delete the `addIArgument` line

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/tensorops/EmbeddingLookup.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    public EmbeddingLookup(@NonNull INDArray in, @NonNull INDArray indices, INDArray output) {
+        super("embedding_lookup", new INDArray[]{in, indices}, new INDArray[]{output});
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/tensorops/EmbeddingLookup.java:48]: output may be null - `wrapOrNull(output)`

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/tensorops/EmbeddingLookup.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        addIArgument(indices.toIntVector());
+
+    }
+
+    public EmbeddingLookup(@NonNull INDArray in, INDArray output, @NonNull int... indices) {
+        super("embedding_lookup", new INDArray[]{in, Nd4j.create(indices)}, new INDArray[]{output});
+        addIArgument(indices);
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/tensorops/EmbeddingLookup.java:55]: I've flagged this before: Don't confuse `Nd4j.create(int[])` with `Nd4j.createFromArray(int[])`
They are very different things. Suppose indices = [1,2,3,4,5]
`Nd4j.create(indices)` in this case gives us a floating point rank 5 array with shape [1,2,3,4,5] and 120 elements; `Nd4j.createFromArray(indices)` gives us a rank 1 integer array with 5 elements and values (1,2,3,4,5).
We want the latter here.

Furthermore: we don't need _both_ the indices passed as an INDArray and as IArgs. It's one or the other. Let's stick with Nd4j.createFromArray and remove the addIArgument line

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/tensorops/EmbeddingLookup.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+
+    }
+
+    @Override
+    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
+        Preconditions
+                .checkArgument(dataTypes != null && dataTypes.size() == 2, "Expected exactly 2 input datatypes, got %s", dataTypes);
+        Preconditions.checkArgument(dataTypes.get(0).isFPType(), "Input datatype must be floating point, got %s", dataTypes);
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/tensorops/EmbeddingLookup.java:64]: We should check that the second input is an integer datatype

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/MaximumBp.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
+package org.nd4j.linalg.api.ops.impl.transforms.custom;
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/MaximumBp.java:1]: Copyright header

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/pairwise/arithmetic/bp/MergeAddBp.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
+package org.nd4j.linalg.api.ops.impl.transforms.pairwise.arithmetic.bp;
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/pairwise/arithmetic/bp/MergeAddBp.java:1]: Copyright header

File: nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/LayerOpValidation.java [hunk 11]
```diff
@@ -734,8 +741,8 @@ public void testConv2dBasic() {
                 .isSameMode(false)
                 .build();
 
-        SDVariable out = sd.cnn().conv2d("conv", in, w, b, c);
-        out = sd.nn().tanh("out", out);
+        SDVariable out = sd.f().conv2d(vars, c);
+        out = sd.f().tanh(out);
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/LayerOpValidation.java:745]: I have already flagged this recently - and there's lots of instances of this that you've changed in LayerOpValidation - don't use f()
f() (i.e., DifferentialFunctionFactory) should be considered deprecated and will be removed soon. Use the namespaces instead.

File: nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/TransformOpValidation.java [hunk 12]
```diff
@@ -1986,4 +2049,159 @@ public void testLogSumExp2(){
...
+
+
+                SDVariable out = new ImageResize(sd, inputImage, requestedSize, preserveAspectRatio, antialias, method).outputVariable();
+                OpValidation.validate(new TestCase(sd)
+                        .gradientCheck(false)
+                        .expected("image_resize", checkFunction));
+
+            } catch (Exception e) {
+                System.out.println("Method failed: " + method);
+            }
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/TransformOpValidation.java:2133]: Don't use try/catch here like this - remove the try/catch here entirely.
If an exception is thrown anywhere within the try block: we'll get `Method failed: ...` printed, but the test itself won't actually fail - it will pass even if there's a significant problem.

File: nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/LayerOpValidation.java [hunk 13]
```diff
@@ -1469,6 +1476,41 @@ public void testBiasAdd_nchw_nhwc() {
...
+                .dataFormat("NCHW")
+                .build();
+
+        SDVariable out = sd.cnn.depthWiseConv2d(in, weights, bias, config);
+        SDVariable loss = sd.standardDeviation("loss", out, true);
+        loss.markAsLoss();
+
+        OpValidation.validate(new TestCase(sd)
+                .gradientCheck(true)
+        );
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/LayerOpValidation.java:1509]: Everywhere you use OpValidation - you _must_ check the return value.
```
String err = OpValidation.validate(...);
assertNull(err);
```
Like here: https://github.com/eclipse/deeplearning4j/blob/master/nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/ReductionOpValidation.java#L766-L767

The reason we do it like this: many tests loop over many test cases. If the test fails on the first problem - you won't see the results for the other cases, which can be really important for debugging (for example - is it an edge case? are all test cases failing? only certain conditions?)

For example, there's 56 test cases here:
https://github.com/eclipse/deeplearning4j/blob/master/nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/ReductionOpValidation.java#L634-L737

File: nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/LayerOpValidation.java [hunk 14]
```diff
@@ -1512,19 +1554,24 @@ public void LSTMLayerTestCase1() {
...
         assertArrayEquals(out, outputs.getOutput().eval().shape());
-        assertArrayEquals(hL, outputs.getLastTimeStepOutput().eval().shape());
-        assertArrayEquals(cL, outputs.getLastCellStateOutput().eval().shape());
+        assertArrayEquals(hL, outputs.getLastOutput().eval().shape());
+        assertArrayEquals(cL, outputs.getLastState().eval().shape());
+
+        sd.setLossVariables(outputs.getLastOutput(),outputs.getLastTimeStepOutput(),outputs.getTimeSeriesOutput());
+        OpValidation.validate(new TestCase(sd)
+                .gradientCheck(true)
+        );
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/LayerOpValidation.java:1575]: As per previous comment: check return value. I won't flag every one with a review comment, so scan through github PR files section to find all the other instances.

**Review by AlexDBlack — CHANGES_REQUESTED** [review]

**Comment by AlexDBlack** [comment]
Closing in favor of: https://github.com/KonduitAI/deeplearning4j/pull/378
